### PR TITLE
Add data science

### PR DESCRIPTION
**Pull request by sriking1783**
Pull request for data science

### CODE REVIEW

File: data_science/calculator.rb [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  def p_value
+    [p_value_index, next_higher_value]
+  end
+
+  def winner
+    conversions = {}
+    @experiment.cohorts.each do |cohort|
+      conversions[cohort.name] = cohort.conversion_rate
+    end
+    conversions.find { |_cohort, conv| conv == conversions.values.max }.first
```
Comment (zsyed1) [inline data_science/calculator.rb]: FYI Using `each_with_object` will save you a line of code (i.e. line 59):
```@experiment.cohorts.each_with_object({}) do |cohort, conversions|
  conversions[cohort.name] = cohort.conversion_rate
end```

Comment (zsyed1) [inline data_science/calculator.rb]: Sorry, multi-line code never seems to work for me in these comment fields.

File: data_science/cohort.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  def initialize(data)
+    @name = data.keys.first
+    @visits = data[@name]["total_visits"]
+    @conversions = data[@name]["conversions"]
+  end
+
+  def conversion_rate
+    return 0 unless visits.nonzero?
+    (conversions.to_f / visits).round(4)
```
Comment (zsyed1) [inline data_science/cohort.rb]: Line 12 is kind of like a double negative. `return 0 if visits.zero?` might be better.

File: data_science/cohort.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  def standard_error
+    Math.sqrt((conversion_rate * (1 - conversion_rate)) / visits).round(4)
+  end
+
+  def expected_conversion_rate
+    sample_mean = conversion_rate * 100
+    { max: (sample_mean + standard_mean).round(2),
+      min: (sample_mean - standard_mean).round(2) }
+  end
```
Comment (zsyed1) [inline data_science/cohort.rb]: Pretty!

File: data_science/experiment.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  def initialize(data)
+    @cohorts = data.map { |key, value| Cohort.new(key => value) }
+  end
+
+  def visits_for_experiment
+    @cohorts.each_with_object({}) do |cohort, visits|
+      visits[cohort.name] = cohort.visits
+    end
+  end
```
Comment (zsyed1) [inline data_science/experiment.rb]: Excellent! So you can ignore my comment above about `each_with_object` - you clearly already know about this.

File: data_science/spec/spec_helper.rb [hunk 5]
```diff
@@ -0,0 +1,4 @@
+require_relative '../json_parser'
+require_relative '../experiment'
+require_relative '../calculator'
+require_relative '../cohort'
```
Comment (zsyed1) [inline data_science/spec/spec_helper.rb]: Very nice overall! This is closed.

File: data_science/spec/calculator_spec.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
+require 'spec_helper'
+
+describe Calculator do
+  let(:calculator) { described_class.new("test_data.json") }
+
+  it "calculates total conversions experiment" do
+    total = calculator.chi_square
+    expect(total).to eq(0.07)
+  end
+  it "calculates the winner to be A" do
```
Comment (djkotowski) [inline data_science/spec/calculator_spec.rb]: It shouldn't always calculate the winner to be 'A'.  That's just this test case.

File: data_science/spec/calculator_spec.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    conversions = calculator.expected_conversions("A")
+    expect(conversions).to eq(3.75)
+  end
+
+  it "calculates the expected failures of the cohort" do
+    failures = calculator.expected_failures("A")
+    expect(failures).to eq(2.25)
+  end
+
+  it "calculates the p_value is significant" do
```
Comment (djkotowski) [inline data_science/spec/calculator_spec.rb]: Ditto on the above comment.

File: data_science/spec/calculator_spec.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    pvalue = calculator.p_value
+    expect(pvalue).to eq([25, 25])
+  end
+  it "calculates the degrees of freedom is significant" do
```
Comment (djkotowski) [inline data_science/spec/calculator_spec.rb]: Ditto.

**Comment by djkotowski** [comment]
Just a few comments, but you're done with this one.  I like the short, pretty methods and your test coverage.

**Pull request closed without merging**